### ISSUE TITLE
[Potential Bug] run_model reuses the same inputs across accum_steps instead of fetching a new batch each step

### ISSUE THREAD

**Issue by Lucenova** [baaivision/URSA#10]
In `run_model`, the following line is executed inside the `accum_steps` loop:

```python
inputs = inputs if inputs else self.train_dataloader.next()[0]
```

From my understanding, once `inputs` becomes non-empty after the first iteration, the same inputs will be reused for all remaining accumulation steps. As a result, when `accum_steps > 1`, the model may repeatedly train on the same batch instead of fetching a new batch for each accumulation step.

I was wondering whether this is the intended behavior. If gradient accumulation is expected to use different micro-batches across accumulation steps, this might potentially be a bug.

I may be misunderstanding the design here, so I would really appreciate any clarification. Thanks in advance!

**Comment by Lucenova** [comment]
<img width="1984" height="1646" alt="Image" src="https://github.com/user-attachments/assets/337ceb6f-691f-49bd-8464-6fd372532231" />

**Comment by PhyscalX** [comment]
Hi, @Lucenova 

We use ``data_every`` to control the dataloader behavior.
In our [UDM-GRPO](https://github.com/Yovecent/UDM-GRPO/blob/d1bec49f4500873606f8345d81692143de059891/diffnext/configs/geneval_grpo/ursa_1.7b_ibq512.yaml#L59) development, set ``data_every=2`` could match the Flow-GRPO's mixed policy update, 
i.e.,  update policy model with 50% samples in a batch, and the next 50% samples will be off-policy.

**Comment by Lucenova** [comment]
> Hi, [@Lucenova](https://github.com/Lucenova)
> 
> We use `data_every` to control the dataloader behavior. In our [UDM-GRPO](https://github.com/Yovecent/UDM-GRPO/blob/d1bec49f4500873606f8345d81692143de059891/diffnext/configs/geneval_grpo/ursa_1.7b_ibq512.yaml#L59) development, set `data_every=2` could match the Flow-GRPO's mixed policy update, i.e., update policy model with 50% samples in a batch, and the next 50% samples will be off-policy.

@PhyscalX Thanks for the clarification — that makes sense for the UDM-GRPO setting and the mixed-policy update behavior.

From my understanding, in the original GRPO-style setup, reusing the same inputs across `accum_steps` is actually intentional, because the same samples are expected to be evaluated under both the old and updated policies. In that context, the current behavior is reasonable and aligns with the mixed-policy semantics.

However, I think the behavior becomes different for standard T2I diffusion training.

In T2I training, there is usually no on-policy / off-policy distinction. The purpose of gradient accumulation is mainly to simulate a larger effective batch size by processing different micro-batches across accumulation steps.

If all accumulation steps reuse the exact same inputs, then the effective batch size does not actually increase — it becomes closer to repeatedly scaling the same loss rather than accumulating gradients from different samples. In that sense, the current logic may be suboptimal (or effectively a bug) for standard T2I training workflows, even though it works correctly for GRPO-style training.

One possible solution might be to explicitly separate these two semantics through something like `data_every`:

* GRPO / mixed-policy training:
  reuse the same inputs across accumulation steps
* Standard T2I training:
  fetch new micro-batches for each accumulation step

Conceptually something like:

```python
for step_idx in range(accum_steps):
    if step_idx == 0:
        if not inputs:
            inputs = self.train_dataloader.next()[0]
    else:
        if self.config.training.get("data_every", -1) >= 1:
            pass  # reuse inputs for GRPO-style mixed-policy update
        else:
            inputs = self.train_dataloader.next()[0]  # fetch new data for T2I
```

That said, if this branch is intended purely for T2I training and not for future GRPO reuse, then the current modification (fetching new data per accumulation step) probably is the cleanest solution.

Thanks again for the explanation — it helped clarify the original design intent.

**Comment by PhyscalX** [comment]
Following the intention of training loop, we process ``data_every`` [[HERE](https://github.com/baaivision/URSA/blob/aa723b212cf4ff4850a9ea6a30d2f3336d2c80c9/diffnext/engine/train_engine.py#L179)].
The default value to ``inputs`` in the ``run_model(...)`` was kept as a legacy interface.

**Comment by Lucenova** [comment]
@PhyscalX Thanks for the clarification — I think I understand the current design better now.

From my understanding, `data_every` controls the data refresh behavior at the training-loop / global-step level, while the internal `accum_steps` inside `run_model(...)` are treated as part of the same optimization step.

In that case, during one gradient accumulation cycle, all accumulation sub-steps intentionally reuse the same batch instead of fetching new micro-batches. Is my understanding correct?

I originally interpreted gradient accumulation in the more standard T2I sense, where each accumulation step usually consumes different micro-batches to simulate a larger effective batch size. So I wanted to confirm whether the current behavior is intentionally designed for the GRPO-style training semantics rather than conventional diffusion training semantics.

Thanks again for the detailed explanation.

**Comment by PhyscalX** [comment]
Your understanding is incorrect.
1. By default, we set ``data_every=-1``, which follows the legacy behavior where the dataloader is iterated at every sub-step.

2. When ``data_every!=-1``, the optimizer is stepped ``data_every`` times while the dataloader is iterated once, finishing ``data_every*gradient_accumulation_steps`` micro batches.

**Comment by Lucenova** [comment]
> 1. By default, we set `data_every=-1`, which follows the legacy behavior where the dataloader is iterated at every sub-step.

@PhyscalX Thank you for the clarification — I really appreciate your patience in walking through this.

I want to make sure I'm not misreading the code, because tracing through the original `run_model` I keep getting a different behavior than the one you described in point (1). Could you help me see where my reading is wrong?

## My trace of `data_every=-1` with `accum_steps=4`

In `train_loop`, with `data_every=-1` the outer `if` branch never fires, so `inputs` stays as the initial `{}` and is passed to `run_step` unchanged on every iteration:

```python
data_every, inputs = self.config.training.get("data_every", -1), {}
while self.global_step < max_steps:
    if data_every >= 1 and self.global_step % data_every == 0:
        inputs = self.train_dataloader.next()[0]
    stats = self.run_step(inputs, accum_steps)

Inside the original run_model:
for _ in range(accum_steps):
    inputs = inputs if inputs else self.train_dataloader.next()[0]
    outputs, losses = self.model(inputs), []
    ...
```

sub-step | inputs at entry | branch taken | inputs at exit
-- | -- | -- | --
0 | {} (falsy) | fetch new batch A | A
1 | A (truthy) | reuse A | A
2 | A (truthy) | reuse A | A
3 | A (truthy) | reuse A | A

From this trace, it looks like the dataloader is iterated once per global_step (not once per sub-step), and the same batch A is reused across all 4 accumulation sub-steps within that step. The local inputs variable inside run_model becomes non-empty after sub-step 0, so the inputs if inputs else ... guard short-circuits for sub-steps 1–3.

This is what made me suspect that, under data_every=-1 with accum_steps>1, the effective batch size stays equal to per_device_batch_size rather than accum_steps × per_device_batch_size.

**Comment by PhyscalX** [comment]
Your analysis is right. 
The gradient accumulation in our legacy behavior is broken after we integrated ``data_every`` argument.

We will fix this bug as follows:
```python
    def run_model(self, inputs, metrics, accum_steps=1):
        reuse_inputs = True if inputs else False
        for _ in range(accum_steps):
            inputs = inputs if reuse_inputs else self.train_dataloader.next()[0]
            outputs, losses = self.model(inputs), []
```

**Comment by Lucenova** [comment]
> Your analysis is right. The gradient accumulation in our legacy behavior is broken after we integrated `data_every` argument.
> 
> We will fix this bug as follows:
> 
>     def run_model(self, inputs, metrics, accum_steps=1):
>         reuse_inputs = True if inputs else False
>         for _ in range(accum_steps):
>             inputs = inputs if reuse_inputs else self.train_dataloader.next()[0]
>             outputs, losses = self.model(inputs), []

Thank you very much for the clarification and for taking the time to carefully investigate this issue. Really appreciate your patience and detailed explanations throughout the discussion. Thanks again for the quick response and fix!
